astropy/stats/funcs.py

```diff
@@ -766,31 +766,35 @@ def median_absolute_deviation(a, axis=None, ignore_nan=False):
     # See https://github.com/numpy/numpy/issues/7330 why using np.ma.median
     # for normal arrays should not be done (summary: np.ma.median always
     # returns an masked array even if the result should be scalar). (#4658)
-    if isinstance(a, np.ma.MaskedArray):
+    if isinstance(data, np.ma.MaskedArray):
         func = np.ma.median
         if ignore_nan:
-            data = np.ma.masked_invalid(data)
+            data = np.ma.masked_where(np.isnan(data), data)
     elif ignore_nan:
-        if NUMPY_LT_1_10:
+        if NUMPY_LT_1_10 or axis is None:
             # older versions don't have nanmedian, but nanmedian
-            # is generally faster than np.ma.median (at least, that is implied
-            # in
-            # https://github.com/astropy/astropy/pull/5232#issuecomment-239173872)
-            data = np.ma.masked_invalid(data)
+            # may be faster than np.ma.median *if* axis is None
+            # https://gist.github.com/MSeifert04/1551e80e1aa85b48c9191e727c38f9a9
```

Comment:
Not sure but I remember it the other way around with `nanmedian` being faster if the axis is None. I'll check it again later.